lgc/patch/PatchBufferOp.cpp

```diff
@@ -1571,3 +1548,54 @@ Instruction *BufferOpLowering::makeLoop(Value *const loopStart, Value *const loo
 
   return loopCounter;
 }
+
+// =====================================================================================================================
+// Create global pointer access.
+//
+// @param bufferDesc: The buffer descriptor
+// @param offset: The offset on the global memory
+// @param type: The accessed data type
+// @param inst: The instruction to be executed on the buffer
+// @param callback: The callback function to perform the specific global access
+Value *BufferOpLowering::createGlobalPointerAccess(Value *const bufferDesc, Value *const offset, Type *const type,
+                                                   Instruction &inst, const function_ref<Value *(Value *)> callback) {
+  // The 2nd element (NUM_RECORDS) in the buffer descriptor is byte bound.
+  Value *bound = m_builder.CreateExtractElement(bufferDesc, 2);
+  Value *inBound = m_builder.CreateICmpULT(offset, bound);
+
+  // If null descriptor or extended robust buffer access is allowed, we will create a branch to perform normal global
+  // access based on the valid check.
+  Value *isValidAccess = m_builder.getTrue();
+  if (m_pipelineState.getOptions().allowNullDescriptor ||
+      m_pipelineState.getOptions().enableExtendedRobustBufferAccess) {
+    Value *isNonNullDesc = m_builder.getTrue();
+    if (m_pipelineState.getOptions().allowNullDescriptor) {
+      // Check dword2 against 0 for null descriptor
+      isNonNullDesc = m_builder.CreateICmpNE(bound, m_builder.getInt32(0));
+    }
+    Value *isInBound = m_pipelineState.getOptions().enableExtendedRobustBufferAccess ? inBound : m_builder.getTrue();
+    isValidAccess = m_builder.CreateAnd(isNonNullDesc, isInBound);
+  }
+
+  BasicBlock *const origBlock = inst.getParent();
+  Instruction *const terminator = SplitBlockAndInsertIfThen(isValidAccess, &inst, false);
+
+  // Global pointer access
+  m_builder.SetInsertPoint(terminator);
+  Value *baseAddr = getBaseAddressFromBufferDesc(bufferDesc);
+  // NOTE: The offset of out-of-bound overridden as 0 may causes unexpected result when the extended robustness access
+  // is disabled.
+  Value *newOffset = m_builder.CreateSelect(inBound, offset, m_builder.getInt32(0));
```

Comment:
As a further tweak, you could drop this `select` if the bounds check is already implied by the branch on `isValidAccess`.

Reply:
Sure. Thank you.